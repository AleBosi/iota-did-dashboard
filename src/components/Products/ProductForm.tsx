import React, { useState } from "react";
import { Product } from "../../models/product";
import { generateDid } from "../../utils/didUtils";
import { saveItem } from "../../utils/storageHelpers";

interface Props {
  parentProduct?: Product; // se fornito, aggiunge come figlio
  onCreate: (product: Product) => void;
}

const ProductForm: React.FC<Props> = ({ parentProduct, onCreate }) => {
  const [typeId, setTypeId] = useState("");
  const [serial, setSerial] = useState("");
  const [owner, setOwner] = useState("");

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    if (!typeId || !serial) return;

    const product: Product = {
      productId: generateDid(),
      typeId,              // ora funge da nome prodotto
      did: generateDid(),
      serial,
      owner,
      credentials: [],
      children: [],
    };

    saveItem(`Product:${product.productId}`, product);

    // Se ha un padre, aggiungi il prodotto come child
    if (parentProduct) {
      parentProduct.children = parentProduct.children || [];
      parentProduct.children.push(product);
      saveItem(`Product:${parentProduct.productId}`, parentProduct);
    }

    onCreate(product);
    setTypeId("");
    setSerial("");
    setOwner("");
  };

  return (
    <form onSubmit={handleSubmit} className="flex flex-col gap-2 mb-4">
      <input
        placeholder="Nome prodotto"
        value={typeId}
        onChange={e => setTypeId(e.target.value)}
        required
        className="border px-2 py-1 rounded"
      />
      <input
        placeholder="Seriale"
        value={serial}
        onChange={e => setSerial(e.target.value)}
        required
        className="border px-2 py-1 rounded"
      />
      <input
        placeholder="Owner DID"
        value={owner}
        onChange={e => setOwner(e.target.value)}
        className="border px-2 py-1 rounded"
      />
      <button
        type="submit"
        className="bg-blue-500 text-white px-4 py-1 rounded"
      >
        {parentProduct ? "Aggiungi sotto-componente" : "Crea prodotto"}
      </button>
    </form>
  );
};

export default ProductForm;
