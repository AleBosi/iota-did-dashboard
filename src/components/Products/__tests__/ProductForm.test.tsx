import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import ProductForm from "../ProductForm";

describe("ProductForm", () => {
  const mockOnCreate = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("renderizza correttamente tutti i campi e il bottone", () => {
    render(<ProductForm onCreate={mockOnCreate} />);
    expect(screen.getByPlaceholderText(/nome prodotto/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/seriale/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/owner did/i)).toBeInTheDocument();
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  test("non submit senza nome o seriale", async () => {
    render(<ProductForm onCreate={mockOnCreate} />);
    fireEvent.click(screen.getByRole("button"));
    await waitFor(() => {
      expect(mockOnCreate).not.toHaveBeenCalled();
    });
  });

  test("submit chiama onCreate con dati corretti e resetta campi", async () => {
    render(<ProductForm onCreate={mockOnCreate} />);
    fireEvent.change(screen.getByPlaceholderText(/nome prodotto/i), { target: { value: "Prodotto Test" } });
    fireEvent.change(screen.getByPlaceholderText(/seriale/i), { target: { value: "SN12345" } });
    fireEvent.change(screen.getByPlaceholderText(/owner did/i), { target: { value: "did:iota:evm:owner" } });

    fireEvent.click(screen.getByRole("button"));

    await waitFor(() => {
      expect(mockOnCreate).toHaveBeenCalledTimes(1);
      const productArg = mockOnCreate.mock.calls[0][0];
      expect(productArg.name).toBe("Prodotto Test");
      expect(productArg.serial).toBe("SN12345");
      expect(productArg.owner).toBe("did:iota:evm:owner");
    });

    expect(screen.getByPlaceholderText(/nome prodotto/i)).toHaveValue("");
    expect(screen.getByPlaceholderText(/seriale/i)).toHaveValue("");
    expect(screen.getByPlaceholderText(/owner did/i)).toHaveValue("");
  });

  test("bottone cambia testo se viene passato parentProduct", () => {
    const parentProduct = { 
      id: "p1", 
      did: "did:iota:evm:parent123", 
      name: "Parent", 
      bom: [] 
    };
    render(<ProductForm parentProduct={parentProduct} onCreate={mockOnCreate} />);
    expect(screen.getByRole("button")).toHaveTextContent(/aggiungi sotto-componente/i);
  });

  // Snapshot
  test("matcha lo snapshot", () => {
    const { asFragment } = render(<ProductForm onCreate={mockOnCreate} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
