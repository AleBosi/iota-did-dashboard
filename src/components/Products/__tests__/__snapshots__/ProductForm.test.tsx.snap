// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProductForm matcha lo snapshot 1`] = `
<DocumentFragment>
  <form
    class="flex flex-col gap-2"
  >
    <input
      placeholder="Nome prodotto"
      required=""
      value=""
    />
    <input
      placeholder="Seriale"
      required=""
      value=""
    />
    <input
      placeholder="Owner DID"
      value=""
    />
    <button
      type="submit"
    >
      Crea prodotto
    </button>
  </form>
</DocumentFragment>
`;
